### PR TITLE
perf(l2): static geometry, block-id transfers, and one-kernel H2D flags

### PR DESCRIPTION
**Pull request by OftenDream**
## Summary

Replace per-layer executor-side byte-range expansion with static field geometry and dynamic block IDs. The executor builds consumer-ordered geometry once, stages block pairs per operation, and lets the mapped-Host Triton kernel compute Device/Host offsets. D2H uses one full-geometry launch; supported NVIDIA H2D uses one grid with per-layer completion flags.

### Transfer path

- Reuse one static geometry table across loads and writebacks. Precompute each field's chunk count instead of converting static geometry tensors to Python lists on every launch.
- Bucket `(group, device_block_id, host_block_id)` mappings by cache group. Loads asynchronously commit an indexed workspace; writebacks finish pinned metadata staging before the caller can refill the shared write workspace.
- Resolve transport capability before selecting the consumer wait protocol. NVIDIA mapped-Host H2D uses per-layer flags; unflagged Triton on other vendors and DMA/NPU use per-layer launches and events. Device-resident geometry alone does not imply flag support.
- Limit automatic DMA fallback to unavailable Host-pointer mapping, cached per workspace and buffer binding. Validation, allocation, and kernel-launch errors propagate instead of silently downgrading. Lazy byte-range expansion remains inside the kernel package's DMA boundary.
- Remove the old Triton range kernels and Device range tables. Scheduler/C++ transfer assembly is unchanged.

### Ordering and lifetime

- A flagged launch must be a full-geometry H2D transfer. CTAs traverse consumers in layer order, release-store each ready flag after the layer copy, and signal empty consumers too. Compute uses acquire waits before accessing the corresponding cache fields.
- Record `layer_ready_init_event` on the load stream after resetting flags and CTA counters. Consumers wait for this event before polling flags, so workspace reuse cannot expose a stale `1` from the previous generation.
- Keep full-transfer completion events separate from per-layer readiness: ACKs and workspace reuse wait for the entire transfer. Failed submissions retire outstanding work without a success ACK; failure to establish retirement poisons the executor rather than allowing unsafe reuse.
- Order cache writebacks and zero-only plans after previously submitted model execution using a cross-stream wait. This prevents copies or block resets from racing prior model cache accesses without adding a normal-path CPU synchronization.

Also retains the profile-rendezvous fix: `/stop_profile` uses a CPU tensor on the attn-TP Gloo group instead of a `barrier()` that could allocate a CUDA tensor on the control-plane thread.

### L2 load CPU profile

The previously captured Kimi-K3 profile compares the old per-layer range/launch path with the single-launch path:

| Python submit window | Before | After |
|---|---|---|
| `_start_loading` | ~101 ms (metadata preparation + 93 launches) | ~3.6 ms |

**Before**

L2 load _start_loading before

**After**

L2 load _start_loading after

These are CPU enqueue measurements from the earlier profile, not a fresh measurement of the latest refactor or an end-to-end TTFT/throughput claim. Actual layerwise overlap depends on transfer length and SM contention.

Co-authored-by: longcat-ia-team <312376714+longcat-ia-team@users.noreply.github.com>

### CODE REVIEW

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py [hunk 1]
```diff
@@ -166,11 +545,230 @@ def _triton_is_unavailable(error: Exception) -> bool:
...
+        )
+        _publish_dma_flags()
+
+    if backend == "dma":
+        transfer_dma()
+        return
+
+    global _mapped_host_triton_available
+    mapped_host_candidate = device_buffers[0].device.type != "npu"
+    if mapped_host_candidate and _mapped_host_triton_available is not False:
```
[the remark this answers is not in the document]
Comment (OftenDream) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py]: Fixed in f5a8017a.

`layer_ready_ptx_supported()` is `current_platform().is_nvidia`. The executor only prepares flags and launches the one-grid path on NVIDIA. Other vendors keep per-layer `transfer_cache_blocks` plus Events, so `fence.acq_rel.gpu` / `st.release.gpu` / `ld.acquire.gpu` are never compiled. `backend="triton"` with flags on a non-NVIDIA device raises; `auto` DMA-fills flags if a caller still passes them.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py [hunk 2]
```diff
@@ -79,83 +297,248 @@ def bind_addresses(
...
+                capacity = max(capacity, self._layer_ready_flags.shape[0] * 2)
+            self._layer_ready_flags = torch.zeros(
+                capacity, dtype=torch.int32, device=device
+            )
+            self._layer_cta_counts = torch.zeros(
+                capacity, dtype=torch.int32, device=device
+            )
+        else:
+            self._layer_ready_flags[:num_layers].zero_()
+            self._layer_cta_counts[:num_layers].zero_()
```
[the remark this answers is not in the document]
Comment (OftenDream) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py:674]: Fixed in d7d45ef6, with the capability refactor and stronger reuse regression coverage in a2620f20.

The stale-`1` race was real: retiring the previous transfer does not order the next asynchronous flag reset against the compute stream. The current sequence is:

1. On `load_stream`, reset the ready flags and CTA counters, then record `layer_ready_init_event` for each active target/draft tracker.
2. In `wait_for_layer()`, the consumer stream waits for that initialization event **before** launching the acquire-polling kernel. It therefore cannot accept a ready flag left by the previous generation.
3. The transfer release-stores each layer's ready flag only after copying that layer. A separate full-transfer finish event still controls ACK publication and workspace reuse; layer readiness is not treated as metadata retirement.

This adds a stream dependency, not a CPU synchronization or a wait for the entire H2D transfer before layer 0 can run. Transport capability is now resolved before publishing the wait protocol; non-flag paths retain per-layer events.

The GPU regression `test_layered_h2d_reuses_workspaces_after_delayed_flag_reset` runs seven generations across three workspaces, deliberately delays reset on the load stream, restores different Host blocks into the same destination, and snapshots each layer on the consumer stream immediately after its wait. All snapshots match their current-generation payload, including after workspace wraparound; the empty-consumer case is covered as well. The host-transfer suite passes all 56 tests, and the runtime/layerwise suites pass 19 tests plus 8 subtests.

Separately, 40f55d6d orders cache writeback/reset operations after prior model execution. That addresses a different cross-stream dependency and is not a substitute for this flag-initialization fence.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py [hunk 3]
```diff
@@ -24,153 +24,830 @@
...
+        self._num_layer_ready = 0
+        self._backend_key = None
+        self._mode: HostTransferMode | None = None
+
+    def prepare_backend(
+        self,
+        device_buffers: Sequence[torch.Tensor],
+        host_buffer: torch.Tensor,
+        *,
+        backend: Literal["auto", "triton", "dma"] = "auto",
```
[the remark this answers is not in the document]
Comment (OftenDream) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py]: Fixed in ebf52a5a. Removed the new transfer-helper defaults for backend selection, staging synchronization, grid cap, and layer-completion metadata, and updated all call sites to pass them explicitly. The runtime still deliberately selects auto where configured; omitting the argument now fails at the API boundary. Added a signature regression test for the public Host transfer helpers.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py [hunk 4]
```diff
@@ -24,153 +24,830 @@
...
+        raise ValueError(
+            f"geometry row {row_index} field payload lies outside host cache block"
+        )
+    if row_packing * host_cache_block_bytes > host_lcm_block_bytes:
+        raise ValueError(
+            f"geometry row {row_index} packed group lies outside host LCM block"
+        )
+
+
+def build_host_transfer_geometry(
```
[the remark this answers is not in the document]
Comment (OftenDream) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py:229]: Fixed in ebf52a5a. The builder docstring now documents every argument, the eight-column field-row schema, byte units, zero-based geometry indices versus one-based cache block IDs, layer-slice coverage and empty consumers, capacity bounds, and the returned HostTransferGeometry with initially unbound Device tables.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py [hunk 5]
```diff
@@ -24,153 +24,860 @@
...
 from tokenspeed_kernel.platform import current_platform
 
-_mapped_host_triton_available: bool | None = None
+
+@dataclass(frozen=True, slots=True)
+class HostTransferMode:
+    """Resolved transport and completion protocol for one buffer binding."""
+
+    backend: Literal["triton", "dma"]
+    layer_ready: bool = False
```
[the remark this answers is not in the document]
Comment (OftenDream) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/kvcache/host_transfer.py]: Fixed in aeb7afc5. Removed the default from HostTransferMode.layer_ready and made every construction pass layer_ready explicitly, including both DMA paths and the Triton capability-dependent path. The existing explicit-argument signature regression now also covers HostTransferMode. You are right that the earlier default-removal change missed this dataclass field; the prior claim was incomplete.
